### PR TITLE
Bug 1121403 — Migration script for room participants

### PR DESCRIPTION
**Pull request by ametaireau**
https://bugzilla.mozilla.org/show_bug.cgi?id=1121403

### CODE REVIEW

File: tools/migrate_1121403_roomparticipants.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+    client.keys('roomparticipants.*', function(err, keys) {
+      if (err) throw err;
+      async.each(keys, function(key, done) {
+        var roomToken = key.split('.')[1];
+        // Access the key using the old format.
+        client.hgetall(key, function(err, participants){
+          // If we have an error, it means we have the right format already.
+          // Skip to the next key.
+          if (err) return done();
```
Comment (Natim) [inline tools/migrate_1121403_roomparticipants.js]: Good catch.

File: tools/migrate_1121403_roomparticipants.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+              multi.psetex(
+                'roomparticipant.' + roomToken + '.' + participant.hawkIdHmac,
+                ttl,
+                JSON.stringify(participant)
+              );
+              multi.sadd('roomparticipants.' + roomToken, participant.hawkIdHmac);
+              multi.pexpire('roomparticipants.' + roomToken, ttl);
+              ok();
+            }, function(err) {
+              multi.exec(done);
```
Comment (Natim) [inline tools/migrate_1121403_roomparticipants.js]: Awesome usage of multi.

File: tools/migrate_1121403_roomparticipants.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            });
+          });
+        })
+      }, callback);
+    })
+  }
+}
+
+if (require.main === module) {
+  main(function() {
```
Comment (Natim) [inline tools/migrate_1121403_roomparticipants.js]: Man for sure you became a Javascript programmer :+1:

File: test/tools_test.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        userMac: 'natim mac',
+        account: 'natim account'
+      }
+    };
+
+    beforeEach(function(done) {
+      var multi = storage._client.multi();
+      roomTokens.forEach(function(token) {
+        Object.keys(participants).forEach(function(id) {
+          var participant = JSON.parse(JSON.stringify(participants[id]));
```
Comment (leplatrem) [inline test/tools_test.js]: Just curious: why you had to do that? Deep copy?

Comment (Natim) [inline test/tools_test.js]: > Deep copy?

Yes indeed.

File: tools/migrate_1121403_roomparticipants.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+ * Migrate the room participants from the old format to the new one.
+ *
+ * Previously, a "roomparticipants.{roomToken}" key was holding the list of
+ * participants, wich their value. The Redis data-type was a hash.
```
Comment (leplatrem) [inline tools/migrate_1121403_roomparticipants.js]: *with*

File: tools/migrate_1121403_roomparticipants.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+ *
+ * Bug 1121403 changed that in order to have expiration events on the
+ * participants; the data type is now a set, refering to all the participants
+ * which are stored in independent keys.
```
Comment (leplatrem) [inline tools/migrate_1121403_roomparticipants.js]: or something like *to all participants ids. Particpants details are now stored with their own independent key.*

File: tools/migrate_1121403_roomparticipants.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+      async.each(keys, function(key, done) {
+        var roomToken = key.split('.')[1];
+        // Access the key using the old format.
+        client.hgetall(key, function(err, participants){
+          // If we have an error, it means we have the right format already.
+          // Skip to the next key.
+          if (err) return done();
+
+          // Delete the hash key since we want to replace it with a set.
+          client.del(key, function(err) {
```
Comment (leplatrem) [inline tools/migrate_1121403_roomparticipants.js]: Shouldn't we first set the roomparticipant details and then replace the key ?

File: tools/migrate_1121403_roomparticipants.js (hunk 3 above)
Comment (leplatrem) [inline tools/migrate_1121403_roomparticipants.js]: :clap: :D

File: tools/migrate_1121403_roomparticipants.js (hunk 7 above)
Comment (ametaireau) [inline tools/migrate_1121403_roomparticipants.js]: Not possible since it's using the same key.

**Pull request merged**